### ISSUE TITLE
Evaluate mode by default?

### ISSUE THREAD

**Issue by Saulzar** [torch/nn#863]
I've been running into a subtle issue with BatchNormalization, namely the fact that it's in training mode by default.

Here's what I've been doing - and I've seen it in other places, I've been passing dummy data through the network in order to compute the sizes at each layer (because I'm lazy), usually an uninitialized Tensor with batch size of 1. Now if BatchNormalization is involved in the network, it 'learns' from that uninitialised data, which subsequently causes the network occasionally to fail to train properly.

At least, that is what I think is happening, because if I'm careful to always use evaluation mode for that purpose, all is well. So, for that reason I think evaluation mode should be default.

**Comment by apaszke** [comment]
Problem is, all modules are initialised in training mode, so this would require changing it all over the place and would be a breaking change.

**Comment by Saulzar** [comment]
Maybe this warning could appear in a prominent place? It's quite a subtle
thing to track down why your network fails to train with BN.

At least if the network is in evaluation mode and you try to
updateGradInput you get a fairly clear error.
